include/onnxruntime/core/framework/tensor.h

```diff
@@ -50,6 +50,61 @@ using BufferNakedPtr = void*;
 #pragma GCC diagnostic ignored "-Wnull-dereference"
 #endif
 #endif
+
+////////////////////////////////////////////////////////////////////////
+/// This class holds optional Meta data describing sparse format
+// when Tensor instance holds data in its sparse format.
+/// When Tensor holds sparse data then:
+/// - it holds only non-zero values in its buffer
+/// - its shape is the shape or rather size of the data in the buffer.
+/// - SparseMetadata must return a non nullptr const pointer to an instance of SparseMetadata
+///
+/// and SparseMetadata contains the enumeration member that indicates its format.
+class SparseMetadata {
+ public:
+  virtual ~SparseMetadata() = default;
+  /// <summary>
+  /// Only one class
+  /// </summary>
+  enum SparseFormat {
+    kNVidiaTwoByFour = 1
+  };
+  SparseFormat Format() const { return format_; }
+  /// <summary>
+  /// Safe way to obtain a ptr to a concrete derived class
+  /// </summary>
+  /// <typeparam name="T"></typeparam>
+  /// <returns></returns>
+  template <class T>
+  const T* Get() const;
+
+ protected:
+  explicit SparseMetadata(SparseFormat format) noexcept : format_(format) {}
+
+ private:
+  SparseFormat format_;
+};
+
+/// <summary>
+/// The only supported type for now that does not require any
+/// additional metadata as the data format 2:4 is really dense.
+/// So Tensor itself would hold the data as usual.
+/// </summary>
+class SparseNVidiaTwoByFour : public SparseMetadata {
+ public:
+  SparseNVidiaTwoByFour() : SparseMetadata(kNVidiaTwoByFour) {}
+};
+
+/// <summary>
+/// Specialize for concrete types.
+/// </summary>
+/// <returns></returns>
+template <>
+inline const SparseNVidiaTwoByFour* SparseMetadata::Get<SparseNVidiaTwoByFour>() const {
```

Comment:
Even for simple types like these looks like we'll end up paying the cost of a vptr. I guess we can't avoid it to accommodate extensions.

Reply:
The design choices I considered:

- Inherit from Tensor. The price is a vptr in Tensor due to a virtual __dtor. But we would need yet another major type which do not need in this case and we already have SparseTensor.
- The presented choice is inline with what @skottmckay has proposed and has an extra member pointer which is unused most of the time. It allows as to add a format dependent metadata. The metadata does not need to actually store data just some indices and offsets. We can store everything still in the Tensor buffer but partition it according to the format we choose for the initializer.